tests/testthat/test-conversion.R

```diff
@@ -255,3 +255,151 @@ test_that("as_adjacency_matrix() works -- dense + not both", {
+  }
+
+  g <- sample_gnp(50, 4 / 50, directed = TRUE)
+  el1 <- as_adj_edge_list(g, mode = "out")
+  el2 <- as_adj_edge_list(g, mode = "in")
+  for (i in 1:vcount(g)) {
+    a <- E(g)[.from(i)]
+    expect_equal(length(a), length(el1[[i]]), ignore_attr = TRUE)
+    expect_equal(sort(el1[[i]]), sort(a), ignore_attr = TRUE)
+  }
+  for (i in 1:vcount(g)) {
+    a <- E(g)[.to(i)]
+    expect_equal(length(a), length(el2[[i]]), ignore_attr = TRUE)
+    expect_equal(sort(el2[[i]]), sort(a), ignore_attr = TRUE)
+  }
+})
+
+test_that("graph_from_graphnel() works", {
+  skip_if_not_installed("graph")
+  suppressPackageStartupMessages(library(graph, warn.conflicts = FALSE))
+
+  g <- sample_gnp(100, 5 / 100)
+  N <- as_graphnel(g)
+  g2 <- graph_from_graphnel(N)
+  gi <- graph.isomorphic.vf2(g, g2)
+  expect_true(gi$iso)
+  expect_equal(gi$map12, 1:vcount(g))
+  expect_equal(gi$map21, 1:vcount(g))
+
+  ## Attributes
+
+  V(g)$name <- as.character(vcount(g):1)
+  E(g)$weight <- sample(1:10, ecount(g), replace = TRUE)
+  g$name <- "Foobar"
+
+  N <- as_graphnel(g)
+  g2 <- graph_from_graphnel(N)
+  expect_isomorphic(g, g2)
+  expect_equal(V(g)$name, V(g2)$name)
+
+  A <- as_adjacency_matrix(g, attr = "weight", sparse = FALSE)
+  A2 <- as_adjacency_matrix(g2, attr = "weight", sparse = FALSE)
+  expect_equal(A, A)
+  expect_equal(g$name, g2$name)
+})
+
+test_that("as_graphnel() does not duplicate loop edges", {
+  skip_if_not_installed("graph")
+
+  mat <- matrix(c(1, 0.5, 0.5, 0), nrow = 2)
+  dimnames(mat) <- list(c("A", "B"), c("A", "B"))
+
+  igr <- graph_from_adjacency_matrix(mat, mode = "undirected", weighted = TRUE)
+
+  grNEL <- as_graphnel(igr)
+  expect_equal(graph::edgeL(grNEL)$A$edges, c(1, 2))
+})
+
+
+test_that("as_long_data_frame() works correctly with and without names", {
+  expect_snapshot({
+    ring <- make_ring(3)
+    as_long_data_frame(ring)
+
+    V(ring)$name <- letters[1:3]
+    as_long_data_frame(ring)
+
+    V(ring)$score <- LETTERS[1:3]
+    as_long_data_frame(ring)
+
+    E(ring)$info <- 3:1
+    as_long_data_frame(ring)
+  })
+})
+
+test_that("as_biadjacency_matrix() works -- dense", {
+  I <- matrix(sample(0:1, 35, replace = TRUE, prob = c(3, 1)), ncol = 5)
+  g <- graph_from_biadjacency_matrix(I)
+  I2 <- as_biadjacency_matrix(g)
+  expect_equal(I, I2, ignore_attr = TRUE)
+  expect_identical(rownames(I2), as.character(1:7))
+  expect_identical(colnames(I2), as.character(8:12))
+})
+
+test_that("as_biadjacency_matrix() works -- dense named", {
+  I <- matrix(sample(0:1, 35, replace = TRUE, prob = c(3, 1)), ncol = 5)
+  g <- graph_from_biadjacency_matrix(I)
+  V(g)$name <- letters[1:length(V(g))]
+
+  expect_true(is_named(g))
+
+  I2 <- as_biadjacency_matrix(g)
+  expect_equal(I, I2, ignore_attr = TRUE)
+  expect_identical(rownames(I2), c("a", "b", "c", "d", "e", "f", "g"))
+  expect_identical(colnames(I2), c("h", "i", "j", "k", "l"))
+})
+
+test_that("as_biadjacency_matrix() works -- dense + attribute", {
```

Comment:
@szhorvat does this test make any sense? It's also needed for #1518

Reply:
The C core will not support adjacency matrices that contain other values than numbers. This means that we either need to drop support for character adjacency matrices, or we need to implement this behaviour in pure R while making sure to match the C core.

There are several things that don't make much sense with non-numerical adjacency matrices. For example, what do you do with multi-edges? How do you combine their character properties? For plain adjacency matrices, what do you do with self-loops in undirected graphs (whose value is normally doubled in the matrix)?

I'm sure we had a discussion about this in the past, and even reached out to someone who used character matrices (very few people do). But I don't remember the decision. We need to search the issue tracker.

Reply:
I suggest aiming to remove string matrix support, and not producing characterization tests for it. Issue: #1542